Create configload.Parser struct to abstract away from Viper

mx-psi opened the pull request:
**Description:** 

- Create `configload.Parser` struct to encapsulate Viper.

**Link to tracking Issue:** First step for #2596

**Testing:** Amended unit tests

**Documentation:** Updated documentation to reflect new interface

mx-psi commented on config/config.go:
```diff
@@ -104,6 +104,71 @@ func NewViper() *viper.Viper {
+// toStringMap creates a map[string]interface{} from a Viper object
+// It is equivalent to v.AllSettings() but it maps nil values
+// It is used here because of https://github.com/spf13/viper/issues/819
+func toStringMap(v *viper.Viper) map[string]interface{} {
```
This can be simplified if/once spf13/viper#819 gets resolved (it seems like Bogdan already stumbled on this one 😄). The code is the same as the `AllSettings` method except that it does not skip over `nil` values.

mx-psi replied:
Not used anymore, `cast.ToStringMap` seems to overcome this issue

mx-psi commented on config/config.go:
```diff
@@ -104,6 +104,71 @@ func NewViper() *viper.Viper {
+
+// ToCustomUnmarshaler creates a custom unmarshaler from a Viper unmarshaled.
+// It should not be used by new code.
+func ToCustomUnmarshaler(viperUnmarshaler func(*viper.Viper, interface{}) error) component.CustomUnmarshaler {
```
I wrote this wrapper for convenience to fix existing tests but I don't think it should be used by new components.

bogdandrutu replied:
Maybe not in this PR, but I would like to remove this before we declare this module stable. Maybe we can open an issue and decide how to get rid of this.

mx-psi replied:
Created #2735 to track this

mx-psi commented:
(Note that contrib tests fail because this is a breaking public API change; we can fix this once this and #2597 are dealt with)

bogdandrutu commented on component/component.go:
```diff
@@ -68,17 +66,17 @@ type Factory interface {
 // the configuration loading system will use to unmarshal the config.
 type ConfigUnmarshaler interface {
 	// Unmarshal is a function that un-marshals a viper data into a config struct in a custom way.
```
Please update comment to avoid `viper`.

bogdandrutu commented on component/component.go:
```diff
@@ -68,17 +66,17 @@ type Factory interface {
+// CustomUnmarshaler is a function that un-marshals map data into a config struct
 // in a custom way.
-// componentViperSection *viper.Viper
+// componentViperSection map[string]interface{}
```
s/componentViperSection/componentSection

bogdandrutu commented on component/component.go:
```diff
@@ -68,17 +66,17 @@ type Factory interface {
 	// intoCfg interface{}
 	//   An empty interface wrapping a pointer to the config struct to unmarshal into.
-	Unmarshal(componentViperSection *viper.Viper, intoCfg interface{}) error
+	Unmarshal(componentSection map[string]interface{}, intoCfg interface{}) error
```
What do you think, should we expose maybe a struct `ConfigLoader` (or a better name) where we can start adding functionalities like `UnmarshalExact`/`ToStringMap`. Then users will no longer need to use something like viper (we will use to implement these funcs and fix all the bugs). Then we can kind of control all the implementations and if a bug is discovered fix all of them?

Right now we can just have:

```
type ConfigLoader struct {
  v *viper.Viper
}

func (cl *ConfigLoader) UnmarshalExact(interface) error {
}

func (cl *ConfigLoader) ToStringMap(interface) map[string]interface{} {
}
```

Not sure if this is the right thing, but I want a second opinion from someone who started to work on this :)

mx-psi replied:
I am not sure I understand where `ConfigLoader` would be used. Would it be on the `CustomUnmarshaler` interface, as in
```go
 type CustomUnmarshaler func(componentSection map[string]interface{}, intoCfg interface{}) error
```
?

Would it be passed to the `config.Load` function? Or something else?

mx-psi replied:
🤦 sorry I meant to say that on the code haha. I think that makes sense and it's better than the `map[string]interface{}` thing, I will change it (and see if I can come up with a better name)

mx-psi commented:
@bogdandrutu I gave it a try, can you have a look? Note that I had to move some things around to prevent import cycles and that I am exposing `Viper()` temporarily because of that (to be removed when fixing #2735)

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+
+// ToStringMap creates a map[string]interface{} from a ConfigLoader
+func (l *Loader) ToStringMap() map[string]interface{} {
+	return cast.ToStringMap(l.v)
```
Not sure this will return the expected result, we need to test this.

mx-psi replied:
I added some tests, looks like it doesn't so I brought back the implementation I had initially.

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+// FromViper creates a Loader from a Viper instance
```
We usually add `.` (dot) at the end of the sentence.

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+	return viper.NewWithOptions(viper.KeyDelimiter(ViperDelimiter))
+}
+
+// NewLoader creates a new Loader instanceg
```
s/instanceg/instance

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+	ViperDelimiter = "::"
+)
+
+// Creates a new Viper instance with a different key-delimitor "::" instead of the
+// default ".". This way configs can have keys that contain ".".
```
refer to the const in comments instead of the value.

bogdandrutu reviewed (commented):
Let's start with the initial PR to add `configload` package and move the viper helpers there.

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+// limitations under the License.
+
+// package configload implements the configuration Loader struct.
+package configload
```
Should we have this in `config` package directly? Or this will cause circular dependencies and we cannot have it at this point, if that is the case we can file an issue to investigate if this will be a possibility.

bogdandrutu replied:
Forget about this it will not be possible

mx-psi replied:
I don't think it's currently possible, for example `config` imports `testcomponents`, which imports `exporterhelper`, which imports `configload`. Maybe after some major refactor it would be possible but I don't think it's worth it.

bogdandrutu commented on config/configload/configload.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+// NewLoader creates a new Loader instanceg
+func NewLoader() *Loader {
```
Probably better to use `Parser`?

mx-psi replied:
Done.

mx-psi commented:
I think I addressed all your comments. I brought back the initial `ToStringMap` definition since `cast.ToStringMap` doesn't work (I got confused since it is used in some `config/config.go` code for unrelated stuff)

bogdandrutu commented on component/component.go:
```diff
@@ -67,18 +68,25 @@ type Factory interface {
+type CustomUnmarshaler func(componentSection *configload.Parser, intoCfg interface{}) error
+
+// ToCustomUnmarshaler creates a custom unmarshaler from a Viper unmarshaler.
+func ToCustomUnmarshaler(viperUnmarshaler func(*viper.Viper, interface{}) error) CustomUnmarshaler {
```
Can we split for the moment this PR into 2:
1. All changes except changes in component.go (we keep the old interface for the moment), but we add the Parser and move helpers for viper.
2. Changes in the component.go (reason is maybe we just don't do these changes anymore and jump to add the new unmarshaler directly on the config, and deal that's it).

mx-psi commented:
@bogdandrutu PTAL, I removed the `component.go` and related changes

bogdandrutu reviewed (commented):
We will iterate more in the following PRs.
